File: Day48/Part3/dungeon_switch.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <tuple>

using namespace std;

struct Edge {
    int to;
    int has_switch;
};

int solveDungeonSwitchboard(int n, const vector<int>& colors, const vector<tuple<int, int, int>>& edges_input, int s, int t) {
    vector<vector<Edge>> adj(n);
    for (const auto& e : edges_input) {
        adj[get<0>(e)].push_back({get<1>(e), get<2>(e)});
        adj[get<1>(e)].push_back({get<0>(e), get<2>(e)});
    }

    // dist[u][phase]
    vector<vector<int>> dist(n, vector<int>(3, -1));
    queue<pair<int, int>> q;

    // Assume start is valid or we are already there
    dist[s][0] = 0;
    q.push({s, 0});

    while (!q.empty()) {
        pair<int, int> curr = q.front();
        q.pop();
        int u = curr.first;
        int phase = curr.second;
        int d = dist[u][phase];

        if (u == t) {
            return d;
        }

        for (const auto& edge : adj[u]) {
            int v = edge.to;
            int next_phase = edge.has_switch ? (phase + 1) % 3 : phase;

            // Check entry condition for v
            // Green is 1
            if ((colors[v] + next_phase) % 3 == 1) {
                if (dist[v][next_phase] == -1) {
                    dist[v][next_phase] = d + 1;
                    q.push({v, next_phase});
                }
            }
        }
    }

    return -1;
}

int main() {
    int N = 3;
    vector<int> Colors = {1, 0, 0}; // 0=Red, 1=Green, 2=Blue
    vector<tuple<int, int, int>> Edges = {
        {0, 1, 1},
        {1, 2, 0}
    };
    int S = 0;
    int T = 2;

    int result = solveDungeonSwitchboard(N, Colors, Edges, S, T);
    cout << "Minimum Steps: " << result << endl;

    return 0;
}

File: Day48/Part3/dungeon_switch.py
from collections import deque

def solve_dungeon_switchboard(n, colors, edges, s, t):
    # Build adjacency list
    adj = [[] for _ in range(n)]
    for u, v, has_switch in edges:
        adj[u].append((v, has_switch))
        adj[v].append((u, has_switch))
        
    # BFS State: (u, phase)
    # phase: 0, 1, 2
    # Distance map
    dist = {}
    
    # Initial check: Can we be at S?
    # Problem says "You start at S". Usually implies S is valid initially.
    # But let's check: (Colors[S] + 0) % 3 == 1 (Green).
    # If not, maybe we can't even start? Or maybe start is always safe?
    # "Kamu hanya bisa kamu masuki bila warnanya sedang Green."
    # Usually start is entered at t=0.
    # Let's assume we are already IN S, so we don't check entry condition for S at t=0.
    # But if we return to S, we check.
    
    queue = deque([(s, 0, 0)]) # u, phase, steps
    dist[(s, 0)] = 0
    
    while queue:
        u, phase, steps = queue.popleft()
        
        if u == t:
            return steps
        
        for v, has_switch in adj[u]:
            next_phase = (phase + 1) % 3 if has_switch else phase
            
            # Check if we can enter v
            # Condition: (Colors[v] + next_phase) % 3 == 1 (Green)
            current_color_v = (colors[v] + next_phase) % 3
            
            if current_color_v == 1:
                if (v, next_phase) not in dist:
                    dist[(v, next_phase)] = steps + 1
                    queue.append((v, next_phase, steps + 1))
                    
    return -1

if __name__ == "__main__":
    N = 3
    # 0=Red, 1=Green, 2=Blue
    Colors = [1, 0, 0] 
    Edges = [
        [0, 1, 1], # Switch
        [1, 2, 0]  # Normal
    ]
    S = 0
    T = 2
    
    result = solve_dungeon_switchboard(N, Colors, Edges, S, T)
    print(f"Minimum Steps: {result}")
